sematic/api/endpoints/external_resources.py

```diff
@@ -0,0 +1,93 @@
+from sematic.db.models.external_resource_record import ExternalResourceRecord
+from sematic.db.models.user import User
+from sematic.db.queries import (
+    get_external_resource_record,
+    get_resource_ids_by_root_id,
+    save_external_resource_record,
+    save_run_external_resource_link,
+)
+
+ROOT_ID_KEY = "root_id"
+
+
+# Allow getting a list of external resource ids.
+# Q: Why not just return the actual resources?
+# A: Because we should have the endpoint for getting an external resource
+#    hooked up such that it actually interacts with the external compute
+#    to get the most up-to-date status of the records. It would be too much work
+#    for one service call to have to do that for a *list* of resources.
+# Q: Ok, then why not just return the resources without updating them for this?
+# A: Because it would be weird to have the server return stale states when queried
+#    in a list but up-to-date ones when queried one-by-one.
+@sematic_api.route("/api/v1/external_resources/ids", methods=["GET"])
+@authenticate
+def get_resources_endpoint(user: Optional[User]) -> flask.Response:
+    if set(flask.request.args.keys()) != {ROOT_ID_KEY}:
+        return jsonify_error(
+            f"Can only get external resources using the query key {ROOT_ID_KEY} ",
+            HTTPStatus.BAD_REQUEST,
+        )
+
+    resource_ids = get_resource_ids_by_root_id(flask.request.args[ROOT_ID_KEY])
+    payload = dict(resource_ids=resource_ids)
+    return flask.jsonify(payload)
+
+
+@sematic_api.route("/api/v1/external_resources/<resource_id>", methods=["GET"])
+@authenticate
+def get_resource_endpoint(user: Optional[User], resource_id: str) -> flask.Response:
+    record = get_external_resource_record(resource_id=resource_id)
+    if record is None:
+        return jsonify_error(
+            "No such resource: {}".format(resource_id), HTTPStatus.NOT_FOUND
+        )
+
+    payload = dict(record=record.to_json_encodable())
+
+    return flask.jsonify(payload)
+
+
+@sematic_api.route("/api/v1/external_resources/<resource_id>", methods=["POST"])
+@authenticate
+def save_resource_endpoint(user: Optional[User], resource_id: str) -> flask.Response:
+    if (
+        not flask.request
+        or not flask.request.json
+        or "record" not in flask.request.json
+    ):
+        return jsonify_error("Request should have 'record' key", HTTPStatus.BAD_REQUEST)
+
+    record_json_encodable = flask.request.json["record"]
+    record = ExternalResourceRecord.from_json_encodable(record_json_encodable)
+
+    if record.id != resource_id:
+        return jsonify_error(
+            "Resource id should match serialized resource record",
+            HTTPStatus.BAD_REQUEST,
+        )
+    record = save_external_resource_record(record)
+    payload = dict(record=record.to_json_encodable())
+
+    return flask.jsonify(payload)
+
+
+@sematic_api.route(
+    "/api/v1/external_resources/<resource_id>/linked_run/<run_id>", methods=["POST"]
+)
```

Comment:
This could be
```
POST /api/v1/runs/<run_id>/external_resources
```
with the resource as payload.
The resource can already exist in the DB, the endpoint simply persists an entry in the join table.

Reply:
To me this sounds like just switching "associate a resource with a run" to "associate a run with a resource". Why is one preferable to the other?

Reply:
Because "linked_run" is not a resource name, and hierarchically runs are the parent resource.

Also, if we already know the linked run at the time of creation of the external resource (I think we do), then this endpoint can be consolidated with the above one into a single:
```
POST /api/v1/runs/<id>/external_resources
```
that would persist the resource and the association in a single transaction.

Reply:
I still think viewing a resource as being owned by a run is not necessarily the best mental model in the ultimate case of resources being usable by multiple runs. And in that case, this ordering makes just as much sense. But since both seem valid and at least to start with each resource belongs to exactly one run I will switch to that.

I do think it would be wrong though to require the entire resource payload in the post to `/api/v1/runs/<run_id>/external_resources`, so I will go with this way:
- create/save resource using a dedicated endpoint under `/api/v1/external_resources`, where you post the whole resource contents
- associate a run and a resource using a post to `/api/v1/runs/<id>/external_resources` that includes a list of resource ids in the payload